reef-runtime-mesos/src/main/java/org/apache/reef/runtime/mesos/driver/MesosScheduler.java

```diff
@@ -0,0 +1,501 @@
+import org.apache.mesos.Protos.Filters;
+import org.apache.mesos.Protos.FrameworkInfo;
+import org.apache.mesos.Protos.Offer;
+import org.apache.mesos.Protos.Resource;
+import org.apache.mesos.Protos.Status;
+import org.apache.mesos.Protos.TaskID;
+import org.apache.mesos.Protos.TaskInfo;
+import org.apache.mesos.Protos.Value;
+import org.apache.mesos.Protos.Value.Type;
+import org.apache.mesos.Scheduler;
+import org.apache.mesos.SchedulerDriver;
+
+import javax.inject.Inject;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * MesosScheduler that interacts with MesosMaster and MesosExecutors.
+ */
+final class MesosScheduler implements Scheduler {
+  private static final Logger LOG = Logger.getLogger(MesosScheduler.class.getName());
+  private static final String REEF_TAR = "reef.tar.gz";
+  private static final String RUNTIME_NAME = "MESOS";
+  private static final int MESOS_SLAVE_PORT = 5051; //  Assumes for now that all slaves use port 5051(default)
+
+  private final String reefTarUri;
+  private final REEFFileNames fileNames;
+  private final ClasspathProvider classpath;
+
+  private final REEFEventHandlers reefEventHandlers;
+  private final ExecutorService executorService;
+  private final MesosRemoteManager mesosRemoteManager;
+
+  private final SchedulerDriver mesosMaster;
+  private final Map<String, Offer> offers = new ConcurrentHashMap<>();
+
+  private int outstandingRequestCounter = 0;
+  private final ConcurrentLinkedQueue<ResourceRequestProto> outstandingRequests = new ConcurrentLinkedQueue<>();
+  private final Map<String, ResourceRequestProto> executorIdToLaunchedRequests = new ConcurrentHashMap<>();
+  private final Executors executors;
+
+  @Inject
+  MesosScheduler(final REEFEventHandlers reefEventHandlers,
+                 final MesosRemoteManager mesosRemoteManager,
+                 final @Parameter(MesosMasterIp.class) String master_ip,
+                 final @Parameter(JobIdentifier.class) String jobIdentifier,
+                 final Executors executors,
+                 final REEFFileNames fileNames,
+                 final ClasspathProvider classpath) {
+    this.mesosRemoteManager = mesosRemoteManager;
+    this.reefEventHandlers = reefEventHandlers;
+    this.executors = executors;
+    this.fileNames = fileNames;
+    this.executorService = java.util.concurrent.Executors.newSingleThreadExecutor();
```

Comment:
DId not move it onto a Wake Stage (for some reason, this is shown as an outdated diff)